modem_cellular: add registration status API

emillindq opened the pull request:
This commit implements a network registration status API,
including return of reject cause in case of denial.

bjarki-trackunit commented:
I don't think this should be stored in a string, it should be an enum, and should get its own API, something like 
```
static inline int cellular_get_registration_status(const struct device *dev,
                                                   enum cellular_access_technology *tech,
				                   enum cellular_registration_status *status)
```
If nessesary, the `enum cellular_registration_status` could be a union of RAT specific status enums

emillindq committed:
modem_cellular: add registration status API

This commit implements a network registration status API,
including return of reject cause in case of denial.

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

emillindq commented:
@bjarki-trackunit agree, please take a look!

bjarki-trackunit commented on include/zephyr/drivers/cellular.h:
```diff
@@ -210,6 +229,33 @@ static inline int cellular_get_modem_info(const struct device *dev,
+ * @param dev Cellular network device instance
+ * @param tech Which access technology to get status for
+ * @param status Registration status for given access technology
+ * @param reject_cause 3GPP TS 24.301 Annex A reject cause code, if status indicates denied
```
This value we need to enumerate and document like the status and access tech, we may also want to bundle it with the status to make the API a bit cleaner, something like
```
struct cellular_registration {
        enum cellular_registration_status *status;
        uint8_t *reject_cause;
};
```

bjarki-trackunit replied:
Or even better, we can make the status data a union

enum cellular_registration_status *status;

struct cellular_registration_status_rejected {
        uint8_t reject_cause;
};

union cellular_registration_status_data {
        struct cellular_registration_status_rejected rejected;
};

struct cellular_registration {
        enum cellular_registration_status *status;
        union cellular_registration_status_data status_data;
};

It may be overkill, what do you think?

emillindq replied:
> This value we need to enumerate and document like the status and access tech, we may also want to bundle it with the status to make the API a bit cleaner, something like...

I'm not following you here, do you mean we should create an enum for every single reject cause as showed in https://www.etsi.org/deliver/etsi_ts/124300_124399/124301/15.04.00_60/ts_124301v150400p.pdf (Look at Annex A)? That would be one huge enum

In addition, I do get the bundling of registration status and reject cause like this, I'm buying that.
```
struct cellular_registration {
        enum cellular_registration_status *status;
        uint8_t *reject_cause;
};
```
However I do disagree on two things:
1. It looks like `status` is a pointer to a living variable inside the driver. I think the returned value should be value at the moment of the call, since if the value changes let's say in the middle of an if-statement in the caller, sounds dangerous.
2. In my opinion, structs that require accesses like this `registration->data.denied.reject_cause` quite severely damages the overview and readability of the code, since it's more layers, requiring you to go back and forth to the struct, and is it a pointer or variable, and so on. Sure, helpful in some cases but I think in this case it's overkill, if you ask me :slightly_smiling_face:

MaureenHelm commented:
@bjarki-trackunit ping

bjarki-trackunit commented:
The main concern I have with this API is the high coupling between specific cellular technologies like GSM and E-UTRAN, like the status REGISTRATION_SMS_ONLY_HOME which is only relevant for a few technologies like E-UTRAN.

The reject_cause is also quite vague, one has to look through the 3gpp spec to figure out what it means in case it is relevant for the specific status.

I really want this API to be simple and technology agnostic, to be useful into the future as well, where new states and properties will likely be added, think CGREG, CEREG, ..., and LTE-M1, NB-IOT etc etc

Simplifying the API into only what is general for all cellular networks, like disabled, searching, denied and roaming, would make this APIs utility broader :)

emillindq commented:
@bjarki-trackunit I can agree with that, I've now reduced the scope of this PR to only include the registration statuses which are common between CREG, CGREG and CEREG and removed the reject cause . Please take a look!

tomi-font commented on drivers/modem/modem_cellular.c:
```diff
@@ -350,22 +350,19 @@ static void modem_cellular_chat_on_imsi(struct modem_chat *chat, char **argv, ui
-	bool is_registered;
-
-	is_registered = modem_cellular_is_registered(data);
+	uint8_t registration_status = 0;
```
```suggestion
	enum cellular_registration_status registration_status;
```

bjarki-andreasen reviewed (approved):
Love it!
